atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/VersionedEventStore.java

```diff
@@ -60,41 +61,31 @@ Collection<LockWatchEvent> getEventsBetweenVersionsInclusive(Optional<Long> mayb
     }
 
     LockWatchEvents retentionEvents(Optional<Sequence> earliestSequenceToKeep) {
-        if (eventMap.size() < minEvents) {
-            return LockWatchEvents.builder().build();
+        int numToRetention = eventMap.size() - minEvents;
+        if (numToRetention <= 0) {
+            return LockWatchEvents.empty();
         }
 
-        // Guarantees that we remove some events while still also potentially performing further retention - note
-        // that each call to retentionEventsInternal modifies eventMap.
+        Stream<LockWatchEvent> events = retentionEvents(numToRetention, earliestSequenceToKeep.orElse(MAX_VERSION));
+
+        // Guarantees that we remove some events while still also potentially performing further retention.
+        // Note that consuming elements from retentionEvents stream removes them from eventMap.
         if (eventMap.size() > maxEvents) {
-            List<LockWatchEvent> overMaxSizeEvents = retentionEventsInternal(eventMap.size() - maxEvents, MAX_VERSION);
-            List<LockWatchEvent> restOfEvents =
-                    retentionEventsInternal(eventMap.size() - minEvents, earliestSequenceToKeep.orElse(MAX_VERSION));
-            return ImmutableLockWatchEvents.builder()
-                    .addAllEvents(overMaxSizeEvents)
-                    .addAllEvents(restOfEvents)
-                    .build();
-        } else {
-            return ImmutableLockWatchEvents.builder()
-                    .addAllEvents(retentionEventsInternal(
-                            eventMap.size() - minEvents, earliestSequenceToKeep.orElse(MAX_VERSION)))
-                    .build();
+            Stream<LockWatchEvent> overMaxSizeEvents = retentionEvents(eventMap.size() - maxEvents, MAX_VERSION);
+            return ImmutableLockWatchEvents.of(Stream.concat(overMaxSizeEvents, events)
+                    .collect(Collectors.toCollection(() -> new ArrayList<>(maxEvents))));
         }
+        return ImmutableLockWatchEvents.of(events.collect(Collectors.toCollection(() -> new ArrayList<>(minEvents))));
     }
 
-    private List<LockWatchEvent> retentionEventsInternal(int numToRetention, Sequence maxVersion) {
-        List<LockWatchEvent> events = new ArrayList<>(numToRetention);
-
+    private Stream<LockWatchEvent> retentionEvents(int numToRetention, Sequence maxVersion) {
         // The correctness of this depends upon eventMap's entrySet returning entries in ascending sorted order.
-        eventMap.entrySet().stream()
-                .takeWhile(entry -> entry.getKey().value() < maxVersion.value())
+        return eventMap.headMap(maxVersion).entrySet().stream()
                 .limit(numToRetention)
-                .forEachOrdered(entry -> {
+                .map(entry -> {
                     eventMap.remove(entry.getKey());
```

Comment:
It looks like the original implementation also did this, but this arrangement of removing from the map seems worrying in terms of non-interference, especially since the `entrySet()` contract suggests that we shouldn't edit the map?

Reply:
We can revert this portion if desired.

`eventMap` is a `ConcurrentSkipListMap` whose [entrySet() iterator/stream](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentSkipListMap.html#entrySet--)
is [weakly consistent](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#Weakly). We can safely remove as we traverse the events to retention.

Reply:
I'd prefer to keep this as a stream as it avoid some of the `Object[]` allocations due to intermediate `List` and the semantics are the same